weave-js/src/components/PagePanelComponents/Home/Browse3/charts/extractData.ts

```diff
@@ -0,0 +1,957 @@
+        fieldsMap.set(fieldKey, {
+          key: fieldKey,
+          types: new Set([valueType]),
+          label: `${source === 'input' ? 'Input' : 'Output'}: ${currentPath}`,
+          source,
+          fullPath: currentPath,
+        });
+      }
+    }
+    // If it's an object (not null or array), recursively traverse it
+    else if (valueType === 'object' && value !== null) {
+      extractKeysFromObject(
+        value,
+        source,
+        fieldsMap,
+        currentPath,
+        maxDepth - 1
+      );
+    }
+  });
+}
+
+export function convertSchemaToAxisFields(
+  schema: DynamicFields
+): ChartAxisField[] {
+  const axisFields: ChartAxisField[] = [];
+
+  // Helper function to create axis field with proper render function
+  const createAxisField = (fieldSchema: FieldSchema): ChartAxisField => {
+    const types = Array.from(fieldSchema.types);
+    let chartType: ChartAxisField['type'] = 'string'; // default
+
+    if (types.includes('number')) {
+      chartType = 'number';
+    } else if (types.includes('boolean')) {
+      chartType = 'boolean';
+    } else if (types.includes('string')) {
+      chartType = 'string';
+    }
+
+    // Custom render function based on type
+    let renderFunc = (value: any) => value?.toString() || '';
+
+    if (chartType === 'boolean') {
+      renderFunc = (value: any) => {
+        if (value === true) return 'true';
+        if (value === false) return 'false';
+        if (value === 1) return 'true';
+        if (value === 0) return 'false';
+        return value?.toString() || '';
+      };
+    }
+
+    return {
+      key: fieldSchema.key,
+      label: fieldSchema.label,
+      type: chartType,
+      render: renderFunc,
+    };
+  };
+
+  // Convert input fields
+  schema.inputFields.forEach(fieldSchema => {
+    axisFields.push(createAxisField(fieldSchema));
+  });
+
+  // Convert output fields
+  schema.outputFields.forEach(fieldSchema => {
+    axisFields.push(createAxisField(fieldSchema));
+  });
+
+  // Convert annotation fields
+  schema.annotationFields.forEach(fieldSchema => {
+    axisFields.push(createAxisField(fieldSchema));
+  });
+
+  // Convert score fields
+  schema.scoreFields.forEach(fieldSchema => {
+    axisFields.push(createAxisField(fieldSchema));
+  });
+
+  // Convert reaction fields
+  schema.reactionFields.forEach(fieldSchema => {
+    const axisField = createAxisField(fieldSchema);
+
+    // Special handling for aggregated reactions (notes, reactions)
+    if (
+      fieldSchema.key === 'reactions.notes' ||
+      fieldSchema.key === 'reactions.reactions'
+    ) {
+      axisField.render = (value: any) => {
+        if (Array.isArray(value)) {
+          return value.join(', ');
+        }
+        return value?.toString() || '';
+      };
+    }
+
+    axisFields.push(axisField);
+  });
+
+  return axisFields;
+}
+
+export function getInputOutputFieldValue(
+  extractedData: ExtractedCallData,
+  fieldKey: string
+): any {
```

Comment:
Here, and in many places, I think you want to use `unknown` and not `any`. The reason is because once something is an `any` it can be used anywhere, without type guards. Opposite: `unknown` basically can't be used anywhere until it is type checked. This will require more type checking in consumption code, but that is generally better so that you avoid unexpected types at runtime.